File: src/protocol/conversation.ts
export type MessageRole = "user" | "assistant" | "tool";
export type MessageStatus = "streaming" | "completed" | "failed" | "cancelled";

export interface Conversation {
  id: string;
  title: string;
  createdAt: string;
  updatedAt: string;
}

export interface ConversationMessage {
  id: string;
  runId: string | null;
  role: MessageRole;
  content: string;
  attachments: MessageAttachment[];
  directoryScopes: MessageDirectoryScope[];
  status: MessageStatus;
  sequence: number;
  createdAt: string;
  updatedAt: string;
}

export interface MessageAttachment {
  name: string;
  size: number;
}

export interface MessageDirectoryScope {
  name: string;
}

export interface ConversationSnapshot {
  conversation: Conversation;
  messages: ConversationMessage[];
}

export interface FindRequest {
  id: string;
}

export interface RenameRequest {
  id: string;
  title: string;
}

export interface DeleteRequest {
  id: string;
}

File: src-tauri/src/protocol/conversation.rs
use serde::Serialize;

use super::common::RecordMetadata;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Conversation {
    pub id: String,
    pub title: String,
    pub metadata: RecordMetadata,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Message {
    pub id: String,
    pub role: MessageRole,
    pub content: String,
    pub status: MessageStatus,
    pub sequence: i64,
    pub metadata: RecordMetadata,
}

#[derive(Debug, Clone, Copy, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum MessageRole {
    User,
    Assistant,
}

#[derive(Debug, Clone, Copy, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum MessageStatus {
    Streaming,
    Completed,
    Failed,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Snapshot {
    pub conversation: Conversation,
    pub messages: Vec<Message>,
}
